### PR TITLE
first implementation of integrated multi-agent, multi-target alarming…

### PR DESCRIPTION
**Pull request by wenningerk**
#This commit basically implements the initial proposal in
[http://clusterlabs.org/pipermail/developers/2015-December/000097.html](http://clusterlabs.org/pipermail/developers/2015-December/000097.html) 
The feature is just renamed from notification(s) to alerts(s) to overcome the possible confusion
introduced by pacemaker using notify as well for talking to resource-agents in a completely different context.

e.g.:
`cibadmin --create --xml-text '<configuration> <alerts> <alert id="notify_9" path="/usr/share/pacemaker/tests/pcmk_alert_sample.sh"> <meta_attributes id="meta_9"> <nvpair id="tstamp9" name="tstamp_format" value="%H:%M:%S.%06N"/> </meta_attributes> <instance_attributes id="global_vars_9"> <nvpair id="global_var9_1" name="variable1" value="1"/> <nvpair id="global_var9_2" name="global2" value="1"/> </instance_attributes> <recipient id="recipient_9" value="/tmp/alerts.log"> <instance_attributes id="local_vars_9"> <nvpair id="local_var9_1" name="variable2" value="2"/> <nvpair id="local_var9_2" name="global1" value="overwritten"/> </instance_attributes> </recipient> </alert> </alerts> </configuration>'`

would create the following section in the cib:

```xml 
<configuration> 
  <alerts>
    <alert id="notify_9" path="/usr/share/pacemaker/tests/pcmk_alert_sample.sh">
      <meta_attributes id="meta_9">
        <nvpair id="tstamp9" name="tstamp_format" value="%H:%M:%S.%06N"/>
      </meta_attributes>
      <instance_attributes id="global_vars_9">
        <nvpair id="global_var9_1" name="variable1" value="1"/>
        <nvpair id="global_var9_2" name="global2" value="1"/>
      </instance_attributes>
      <recipient id="recipient_9" value="/tmp/alerts.log">
        <instance_attributes id="local_vars_9">
          <nvpair id="local_var9_1" name="variable2" value="2"/>
          <nvpair id="local_var9_2" name="global1" value="overwritten"/>
        </instance_attributes>
      </recipient>
    </alert>
  </alerts>
</configuration>
```

Till now there is no high-level-tool-support for the feature.
There can be an arbitrary number of alerts and each of the alerts can have an arbitrary number of recipients.
For the full set of properties have a look at .../xml/alerts-2.5.rng

In the discussion in the mailing-list referenced above a drop-in directory for alert-agents was suggested.
Discussion lead to a point where it turned out that the drop-in-directory approach would
need a manual synchronization trigger. (dependency to inotify is unwanted and still might
suffer from races dependent on how the files are created in the drop-in directory)
So it is as well possible to use a high-level-tool like e.g. pcs for feeding the content of a drop-in
directory into the cib - and thus serve as kind of a trigger.

If RHEL7_COMPAT is defined the interface of the previous feature 'integrated-notifications'
configured via the 2 cluster-properties (notification-agent & notification-recipient) is internally 
fed into the new mechanism as long as the new alerts section is not defined in the cib.
Otherwise the alerts section will overrule the previous feature and a warning will be issued.

The interface to the alert-agents is realized as set of environment variables being set accordingly.
It is realized as backward-compatible evolution of the interfaces previously provided with ClusterMon
and 'integrated-notifications'.
To preserve script-compatibility the environment-variables are available prepended CRM_notify as
well as CRM_alert (the implementation is made in a way that is is easy to switch on/off either the
original set with CRM_notify or the new set with CRM_alert without changing the code-logic.)
See the full list of variables at the beginning of .../crmd/notify.c.

For the explanation of variables already available with ClusterMon see 
    http://clusterlabs.org/doc/en-US/Pacemaker/1.1/html/Pacemaker_Explained/s-notification-external.html.
For these which come with the 'integrated-notifications' feature see
    http://blog.clusterlabs.org/blog/2015/reliable-notifications/

On top we have now:
    CRM_notify_node_sequence - a per node sequence which your script might use to sort alerts
    CRM_notify_timestamp - a timestamp created within crmd to bypass delay/jitter coming from
        process spawning & script-execution
    a set of instance_attributes passed 1:1 as environment variables with arbitrary names and
    content to be set as nvpairs as part of the alert-sections and recipient-sections in the cib.
        each alert-section has a set of it's own and this can be amended by additional variables
        put in the recipient-section as seen in the example above.

The timestamp can be configured in a flexible way via the property-string tstamp_format holding a string as defined in the man-page of strftime (+%N as used with date, examples: %i ... seconds since epoche, %H:%M:%S.%06N ... for us accuracy, ...).
 
The commit is not foreseen to be pulled as is, as there are still open points regarding following issues:

- renaming of functions and filenames from notify to alert
- moving the usecond-resolution & format-string time-handling to a common place (common pacemaker time-handling just has seond-resolution and is thus not suitable for the timestamp purpose here)
- having enable/disable per alert-agent (theoretically possible with ClusterMon)
- having node-specific execution of alert-agents (theoretically possible with ClusterMon)
- having possibility to run an alert-agent dependent on a resource/score (theoretically with ClusterMon) 
- and probably more ...

I still chose to provide this commit as it is complete in regards of the cib-interface and the alert-agent-interface so that it can be used as a basis for discussion/implementation of high-level-tools.
And apart from that a pull-request is a good platform for discussion once there is some code available.

Regards,
Klaus

### CODE REVIEW

File: crmd/notify.c [hunk 1]
```diff
@@ -20,120 +20,501 @@
...
+                    continue;
+                }
+                fmt_pos = scanned_pos; /* print till end */
+            }
+        } else {
+            scanned_pos = strlen(format);
+            fmt_pos = scanned_pos; /* print till end */
+        }
+        tmp_fmt_s = strndup(&format[printed_pos], fmt_pos - printed_pos);
+#pragma GCC diagnostic push
```
Comment (jnpkrn) [inline crmd/notify.c]: gcc supports push/pop since 4.6, best to conditionalize, e.g., akin to a recent warning suppression in libqb:

- https://github.com/ClusterLabs/libqb/commit/ea35cfaa616096dd1d770fbda5a735108e1725f8
- https://github.com/ClusterLabs/libqb/commit/e4909329071e3b3805609e38c535e75f69cf9c2f

Can certainly be done in a smoother fashion (also statically per defines injected by the very compiler).

File: xml/crm.dtd [hunk 2]
```diff
@@ -65,7 +65,7 @@ Additionally, num_peers and have_quorum are used during the election process to
  * ccm_transition stores the membership instance from the ConsensusClusterMembership layer.
  * cib_feature_revision is the feature set that this configuration requires
 -->
-<!ELEMENT configuration (crm_config, nodes, resources, constraints)>
+<!ELEMENT configuration (crm_config, nodes, resources, constraints, alarms)>
```
Comment (jnpkrn) [inline xml/crm.dtd]: Would expect the DTD is just the static snapshot of the schema long, long time ago, not suitable for updates any more.

**Comment by kgaillot** [comment]
Haven't had a chance to look at this in detail yet, but it's looking good.

One thing I would change, I think we should use "alert" rather than "alarm". "Alarm" implies either a notification of something wrong, or a scheduled notification; "alert" merely implies a notification of an important event.

File: crmd/cib.c [hunk 3]
```diff
@@ -68,20 +68,39 @@ do_cib_updated(const char *event, xmlNode * msg)
...
     if (format == 1) {
-        if (get_xpath_object
-            ("//" F_CIB_UPDATE_RESULT "//" XML_TAG_DIFF_ADDED "//" XML_CIB_TAG_CRMCONFIG, msg,
-             LOG_TRACE) != NULL) {
+        if ((get_xpath_object(
+                 "//" F_CIB_UPDATE_RESULT "//" XML_TAG_DIFF_ADDED "//" XML_CIB_TAG_CRMCONFIG,
+                 msg, LOG_TRACE) != NULL) ||
+            (get_xpath_object(
+                 "//" F_CIB_UPDATE_RESULT "//" XML_TAG_DIFF_ADDED "//" XML_CIB_TAG_NOTIFICATIONS,
+                 msg, LOG_TRACE) != NULL)) {
```
Comment (kgaillot) [inline crmd/cib.c]: Beware that get_xpath_object() logs a message and dumps the XML if a match is not found (only at trace level, but that can still be distracting when tracing, if it's not relevant). This probably should have already been using xpath_search() instead (since we don't need the result). Also, wherever possible, we should use a single xpath (xpaths can handle "|" for union, "or" and "and", though I'm not sure of the syntax).

File: crmd/cib.c [hunk 4]
```diff
@@ -68,20 +68,39 @@ do_cib_updated(const char *event, xmlNode * msg)
...
-                mainloop_set_trigger(config_read);
-                break;
+
+            if (xpath == NULL) {
+                continue;
             }
+		
+            /* modifying properties */
+            if (!strstr(xpath, "/" XML_TAG_CIB "/" XML_CIB_TAG_CONFIGURATION "/" XML_CIB_TAG_CRMCONFIG "/") &&
+                !strstr(xpath, "/" XML_TAG_CIB "/" XML_CIB_TAG_CONFIGURATION "/" XML_CIB_TAG_NOTIFICATIONS)) {
```
Comment (kgaillot) [inline crmd/cib.c]: FYI, one thing I've started doing recently is defining constants for xpath fragments, e.g.

#define XPATH_CONFIGURATION "/" XML_TAG_CIB "/" XML_CIB_TAG_CONFIGURATION

It definitely makes the code look nicer, but I don't know yet whether it will be easier or harder to follow when I come back to look at it in 6 months.

File: crmd/cib.c [hunk 5]
```diff
@@ -68,20 +68,39 @@ do_cib_updated(const char *event, xmlNode * msg)
...
+                const char *name = NULL;
+
+                /* adding notifications section */
+                if ((strcmp(xpath, "/" XML_TAG_CIB "/" XML_CIB_TAG_CONFIGURATION) != 0) ||
+                    ((section = __xml_first_child(change)) == NULL) ||
+                    ((name = crm_element_name(section)) == NULL) ||
+                    (strcmp(name, XML_CIB_TAG_NOTIFICATIONS) != 0)) {
+                    continue;
+                }
+            }
```
Comment (kgaillot) [inline crmd/cib.c]: What's the purpose of this "if" block? It seems like it just negates the previous change.

File: crmd/control.c [hunk 6]
```diff
@@ -1061,6 +1062,18 @@ config_query_callback(xmlNode * msg, int call_id, int rc, xmlNode * output, void
...
+        sub_call_id = fsa_cib_conn->cmds->query(fsa_cib_conn, 
+            "/" XML_TAG_CIB "/" XML_CIB_TAG_CONFIGURATION
+            "/" XML_CIB_TAG_NOTIFICATIONS "/" XML_CIB_TAG_NOTIFY, NULL,
+            cib_scope_local | cib_xpath);
+
+        fsa_register_cib_callback(sub_call_id, FALSE, NULL,
+                                  notifications_query_callback);
+
+        crm_trace("Querying the CIB for notifications ... call %d", sub_call_id);
+    }
```
Comment (kgaillot) [inline crmd/control.c]: If it's possible, I'd rather expand the crm_read_options() query to search for both crmconfig and alerts in one xpath, and process the results all at once. I'd rather avoid any possible race conditions in two separate queries, and it's nice to save the overhead of two IPC calls.

File: crmd/notify.c [hunk 7]
```diff
@@ -20,120 +20,501 @@
...
 char *notify_script = NULL;
 char *notify_target = NULL;
+GListPtr notify_list = NULL;
+
+typedef struct {
+    char *id;
+    char *path;
+    int timeout;
+    char *tstamp_format;
+    char *info;
```
Comment (kgaillot) [inline crmd/notify.c]: I'm on the fence about "info". What's a compelling use case? Recipient is free-form and can be whatever the script wants, and the id or an XML comment can be used to indicate the script's purpose.

File: crmd/notify.c [hunk 8]
```diff
@@ -20,120 +20,501 @@
...
+
+struct crm_time_us {
+    int years;
+    int months;                 /* Only for durations */
+    int days;
+    int seconds;
+    int offset;                 /* Seconds */
+    bool duration;
+    int useconds;
+};
```
Comment (kgaillot) [inline crmd/notify.c]: Why not just add useconds to crm_time_s? The existing functions could continue to ignore it, and just the new functions would use it. We can add fields to the end of structs without breaking binary compatibility.

File: crmd/notify.c [hunk 9]
```diff
@@ -20,120 +20,501 @@
...
+            .id = notify_script,
+            .path = notify_script,
+            .timeout = CRMD_NOTIFY_DEFAULT_TIMEOUT_MS,
+            .recipient = notify_target
+        };
+        add_dup_notify_list_entry(&entry);
+        crm_notice("Legacy Notifications enabled");
+    } else {
+        crm_notice("Legacy Notifications disabled");
+    }
```
Comment (kgaillot) [inline crmd/notify.c]: I'm thinking the configuration should be either-or: if any new-style alerts are configured, ignore the legacy options. The legacy compatibility is only to preserve the operation of configurations that existed before the new syntax. Once someone switches to the new syntax, any leftover old syntax should be ignored (we could log a message if that's the case).

XML IDs can't contain slashes, so it's probably best to use a static string for id in the legacy case, since there can only be one script configured.

Don't bother logging a message if legacy options aren't in use; the proportion of configurations using them should be tiny, and there's no need to increase the logs of everyone else.

File: crmd/notify.c [hunk 10]
```diff
@@ -20,120 +20,501 @@
...
+            .path = (char *) crm_element_value(notify, XML_NOTIFY_ATTR_PATH),
+            .timeout = crm_get_msec(crm_element_value(notify, 
+                                    XML_NOTIFY_ATTR_TIMEOUT)),
+            .tstamp_format = (char *) crm_element_value(notify, 
+                             XML_NOTIFY_ATTR_TSTAMP_FORMAT),
+            .info = (char *) crm_element_value(notify, XML_NOTIFY_ATTR_INFO)
+        };
+        if (entry.timeout < 0) {
+            entry.timeout = CRMD_NOTIFY_DEFAULT_TIMEOUT_MS;
+        }
```
Comment (kgaillot) [inline crmd/notify.c]: probably <=0, I don't think we should let people disable timeout

File: extra/pcmk_alarm_sample.sh [hunk 11]
```diff
@@ -21,9 +21,14 @@ if [ -z $CRM_notify_version ]; then
     exit 0
 fi
 
+tstamp=`printf "%04d. " "$CRM_notify_node_sequence"`
+if [ ! -z $CRM_notify_timestamp ]; then
+    tstamp="${tstamp} $CRM_notify_timestamp (`date "+%H:%M:%S.%06N"`): " 
+fi
```
Comment (kgaillot) [inline extra/pcmk_alarm_sample.sh]: I can see us adding a variety of sample scripts in the future. Perhaps we should create an extra/alerts directory and move this and pcmk_snmp_helper.sh to it.

File: xml/crm.dtd (hunk 2 above)
Comment (wenningerk) [inline xml/crm.dtd]: From how I got it this defines the absolute anchor-points of the sections below configuration.

**Comment by beekhof** [comment]
Why are we dealing with timestamp formats?
Any reason not to give them an int and let them deal with it?

File: extra/pcmk_alarm_sample.sh (hunk 11 above)
Comment (beekhof) [inline extra/pcmk_alarm_sample.sh]: Agreed

File: crmd/notify.c (hunk 9 above)
Comment (beekhof) [inline crmd/notify.c]: +1

File: crmd/notify.c (hunk 7 above)
Comment (beekhof) [inline crmd/notify.c]: agreed

File: crmd/control.c (hunk 6 above)
Comment (beekhof) [inline crmd/control.c]: +1

File: xml/crm.dtd (hunk 2 above)
Comment (beekhof) [inline xml/crm.dtd]: Right, but anyone using the DTD doesn't get access to the new stuff.
Its deep frozen pending removal.

**Comment by wenningerk** [comment]
Reason for timestamp format was that it should serve as means to order alerts coming from different nodes or being out of order because of scheduling-jitter of scripts running nearly in parallel.
And a resolution of 1s is not enough for that. I guess you share the experience of having logs from
different nodes/applications with 1s resolution timestamps...
Next idea is seconds with us after Epoche but that is quite unusual. 
So my next idea was: "Why not make it versatile so that people can spare further parsing and conversion and can have it as simple as possible and as complex as needed for their purpose."
If you want seconds since Epoche give it an "%i" or "%i.%06N" or ...

File: crmd/notify.c (hunk 8 above)
Comment (wenningerk) [inline crmd/notify.c]: That is of course the goal. The current timestamp-handling is just a quick hack adding what
is needed to notify.c and using the old implementation as helpers and for that you need both
formats unless you want to deal with casting them around and having in mind which version is behind a pointer.

File: crmd/notify.c (hunk 7 above)
Comment (wenningerk) [inline crmd/notify.c]: We support a list of recipients so I thought a field common for all recipients of one script might be handy. In terms of email it could for instance be a subject/part of a subject, a mail-domain, ...
Regarding being used for descriptive purposes I of course thought of XML comments as well but I haven't seen them being used by high-level-tooling  like pcs and it might be handy to have them at hand in the script as well to do some local logging for instance.

File: crmd/cib.c (hunk 5 above)
Comment (wenningerk) [inline crmd/cib.c]: Implying you are referring to the if under "adding notifications section" comment:
In case we are adding the whole section xpath doesn't hold the section-name but it is found
just in the name-property of the 1st child.

File: crmd/cib.c (hunk 4 above)
Comment (wenningerk) [inline crmd/cib.c]: Usually things making the code more compact foster understanding the big picture.
So why not. Question is just if we shouldn't do that globally as a separate commit
(after or before my addition).

File: crmd/control.c (hunk 6 above)
Comment (wenningerk) [inline crmd/control.c]: Didn't want to change too much in the first attempt to create something working to be provided to pcs-development. Also wasn't sure if it would turn out to separate the handling
of that section completely from crm_config.
I also thought that changes wouldn't happen in these parts of the cib that
it would matter performance-wise. And it would complicate the parsing in the next step - right?
Anyway playing with xml-handling is good practice ;-) But please let me tackle that issue
after addressing the things that affect user behaviour.

File: crmd/cib.c (hunk 3 above)
Comment (wenningerk) [inline crmd/cib.c]: You mean replace by a single xpath_search with a combined xpath?

File: crmd/notify.c (hunk 9 above)
Comment (wenningerk) [inline crmd/notify.c]: Good comment. I already fooled myself needing a couple of thoughts why my sequence 
counter was counting up in steps of 2 ;-)
So I go for existence of entries in the new section disabling the legacy feature and if
the legacy-properties are present as well it will result in a warning. The notice that 
the legacy-feature is being used should probably stay.

File: crmd/cib.c (hunk 3 above)
Comment (kgaillot) [inline crmd/cib.c]: Yes

File: crmd/cib.c (hunk 4 above)
Comment (kgaillot) [inline crmd/cib.c]: Either way is fine, separate commit before, or part of this one. I would keep it local to the file (as opposed to putting in a header file)

File: crmd/cib.c (hunk 5 above)
Comment (kgaillot) [inline crmd/cib.c]: OK, I see. Is it impossible for name to be XML_CIB_TAG_CRMCONFIG?

File: crmd/notify.c (hunk 7 above)
Comment (kgaillot) [inline crmd/notify.c]: I can see the usefulness, but I don't feel it adds enough to be worth the extra complexity (one more variable for a user to figure out). If someone wants to use a generic script with multiple clusters, they can create their own local config file for it to read, which gives them the flexibility to decide how they want to do that (format, config management, etc.).

If we did want to do this in pacemaker, I think I'd rather offer a generic name-value pair functionality, and let the user specify what variables they want. I.e. each <alert> could have an <instance_attributes> section. Recipient could even fit into that, but we'd be back to just one recipient (although the script could do something like allow the recipient to be a comma-separated list). For consistency and future flexibility, timeout could even be in a <meta_attributes> section. It sounds like a good bit of extra work and complexity, but it would be consistent with other XML objects. What do you think?

File: xml/crm.dtd (hunk 2 above)
Comment (kgaillot) [inline xml/crm.dtd]: I looked at this more ... crm.dtd and crm-transitional.dtd are essentially pacemaker-0.x.rng. They predate the rng versioning, but they are used similarly by pacemaker code, i.e. for validation and upgrading of CIB syntax (search for "schema" in lib/common/xml.c). So we do need to keep them around for now, and certainly not modify them.

0.6.7 (the last released version to use DTDs) was released in 2008, so perhaps we can think about retiring them, but I don't think it's worth the bother at this point. 2018 would be a nice round 10 years. I'm thinking that might be a good opportunity to bump to 1.2.0, with the bump justified as losing compatibility with pacemaker < 1.0 (and there's probably a good bit of legacy code that could be removed, too).

Unrelated, I noticed that xml.c uses floats to represent XML versions. We need to change that before hitting x.10 ...

File: crmd/cib.c (hunk 5 above)
Comment (wenningerk) [inline crmd/cib.c]: As the basic cluster properties should always be there I guess you would never
have a piece of xml pushed in that actually creates the crm_config-section
So I would say no. Or do you have a bug in mind that would be explained by
that?

File: crmd/notify.c (hunk 7 above)
Comment (wenningerk) [inline crmd/notify.c]: For the first paragraph I consider it rather error-prone to copy around scripts and have multiple versions of them which just differ in a string somewhere - and of course and that is the dangerous part the fixes that happen to just a subset of them. On a production-cluster I would rather distribute generic-scripts in rpms than copy something somewhere in the filesystem.

For the second paragraph - this sounds interesting and I will invest further thinking in a
more generic approach.

Comment (wenningerk) [inline crmd/notify.c]: Having multiple recipients handed over comma separated brings us to a point we've already discussed with multiple ports to collect the status from in ipmi-fence-agent. You loose the
possibility to have a timeout for a single operation observed by pacemaker.

File: crmd/cib.c (hunk 5 above)
Comment (kgaillot) [inline crmd/cib.c]: Nothing in mind, just thinking through the code.

File: crmd/notify.c (hunk 7 above)
Comment (kgaillot) [inline crmd/notify.c]: For the first suggestion, I was thinking that the script itself would stay the same on all hosts, but someone could write it such that it has its own configuration file, which could vary by host. I think that's more useful than a single info parameter.

The second suggestion would let the entire configuration stay within pacemaker.

I can see some value in per-recipient timeouts: the error messages would say which one timed out, and if the number of recipients varied over time, the user wouldn't have to worry about changing the timeout. But that's a relatively small benefit. I'd be open to keeping recipient separate from instance_attributes, but it does seem cleaner and more consistent to keep it in (if going that route). For example, we already have logic for selecting instance attribute sets based on rules, so it would be a little easier to enable rules for alerts in the future, compared to a completely new XML object.

Comment (wenningerk) [inline crmd/notify.c]: A local config-file would be a possibility. But would people spend that effort on their scripts?

The entire config being in pacemaker and visible through the usual tooling throughout the cluster is definitely appealing.

per-recipient timeout benefit I had in mind:
If it hangs for one recipient this would be killed way faster (number of recipients x faster) and further notification wouldn't be blocked. Further on we have parallel notification for all recipients without the script having to implement anything for that. But maybe having problems with it ;-) But that is a different story and would again lead us to serialization of certain streams of alerts.

Any pointer to that instance attribute sets you mentioned?

Comment (kgaillot) [inline crmd/notify.c]: For the schema, see e.g. xml/resources-1.3.rng (element-resource-extra) and xml/nvset-1.3.rng. For the code, see lib/pengine/rules.c:unpack_instance_attributes() and uses of it.

File: xml/crm.dtd [hunk 12]
```diff
@@ -65,7 +65,7 @@ Additionally, num_peers and have_quorum are used during the election process to
  * ccm_transition stores the membership instance from the ConsensusClusterMembership layer.
  * cib_feature_revision is the feature set that this configuration requires
 -->
-<!ELEMENT configuration (crm_config, nodes, resources, constraints)>
+<!ELEMENT configuration (crm_config, nodes, resources, constraints, alerts)>
```
Comment (kgaillot) [inline xml/crm.dtd]: No need to change this. crm.dtd is the original xml schema before we had versioned rng's -- it's only ever used for validating configs that were generated pre-pacemaker-1.0 and have never been updated.

File: crmd/notify.c [hunk 13]
```diff
@@ -20,154 +20,721 @@
...
+notifications_query_callback(xmlNode * msg, int call_id, int rc, 
+                             xmlNode * output, void *user_data)
+{
+    xmlNode *notify = output;
+    notify_entry_t entry;
+
+    free_notify_list();
+
+    if (rc != pcmk_ok) {
+        crm_notice("No optional alerts section in cib");
```
Comment (kgaillot) [inline crmd/notify.c]: I think we can lower this (and the corresponding log below) to info level. In general, notice is to help sysadmins troubleshoot, while info is to help us troubleshoot.

File: crmd/notify.c [hunk 14]
```diff
@@ -20,154 +20,721 @@
...
+                get_envvars_from_cib(recipient, 
+                                     entry.environment_variables, 
+                                     &envvars_added);
+
+            add_dup_notify_list_entry(&entry);
+
+            crm_notice("Notification has recipient: id=%s, value=%s, "
+                       "%d additional environment variables", 
+                       crm_element_value(recipient, XML_NOTIFY_ATTR_REC_ID), 
+                       entry.recipient, envvars_added);
```
Comment (kgaillot) [inline crmd/notify.c]: What's the use case for individual recipients having different environment variables? That sounds like too much potential for confusion to me.

File: xml/alerts-2.5.rng [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+          <optional>
+            <attribute name="timeout"><text/></attribute>
+          </optional>
+          <optional>
+            <!-- format string fed into strftime defaults to %s -->
+            <!-- use format as with strftime + %N as with date -->
+            <attribute name="tstamp_format"><text/></attribute>
+          </optional>
+          <zeroOrMore>
+            <element name="environment_variables">
```
Comment (kgaillot) [inline xml/alerts-2.5.rng]: I'm leaning to using the existing "instance_attributes" XML tag instead of adding a new "environment_variables" tag. The main reason is that users are familiar with instance attributes (in the context of resources) as collections of nvpair elements that are not used by pacemaker but passed directly to the (resource agent) script. It's also used for node attributes, so it helps familiarity. Of course they can still be passed as environment variables in practice.

Similarly, I think it would be consistent to use "meta_attributes" for the timeout, timestamp format, etc., again because users are familiar with those being used by pacemaker itself, and pcs already has commands that define instance attributes and meta-attributes, so the new pcs command could follow a similar approach (e.g. pcs alert create <script> [<options>] [meta <options>] recipient <r> ...).

File: extra/alerts/pcmk_alert_sample.sh [hunk 16]
```diff
@@ -56,12 +61,12 @@ case $CRM_notify_kind in
...
-		echo "Resource operation '${CRM_notify_task}${CRM_notify_interval}' for '${CRM_notify_rsc}' on '${CRM_notify_node}': ${CRM_notify_desc}${CRM_notify_target_rc}" >> ${CRM_notify_recipient}
+		echo "${tstamp}Resource operation '${CRM_notify_task}${CRM_notify_interval}' for '${CRM_notify_rsc}' on '${CRM_notify_node}': ${CRM_notify_desc}${CRM_notify_target_rc}" >> ${CRM_notify_recipient}
 		;;
 	esac
 	;;
     *)
-        echo "Unhandled $CRM_notify_kind notification" >> ${CRM_notify_recipient}
+        echo "${tstamp}Unhandled $CRM_notify_kind notification" >> ${CRM_notify_recipient}
 	env | grep CRM_notify >> ${CRM_notify_recipient}
         ;;
```
Comment (kgaillot) [inline extra/alerts/pcmk_alert_sample.sh]: Great, let's also move pcmk_snmp_helper.sh to the alerts directory (and update the comments to reflect the new configuration).

**Comment by kgaillot** [comment]
Once this is released, we'll need to update the "Receiving Notifications" chapter of Pacemaker Explained.

File: crmd/notify.c [hunk 17]
```diff
@@ -20,154 +20,721 @@
...
+    }
+    return hr_dt;
+}
+
+void
+crm_time_hr_free(crm_time_hr_t * hr_dt)
+{
+    if (!hr_dt) {
         return;
     }
```
Comment (kgaillot) [inline crmd/notify.c]: FYI not necessary, free() ignores NULLs

File: crmd/notify.c [hunk 18]
```diff
@@ -20,154 +20,721 @@
...
+{
+    environment_variable_t *new_entry = 
+        (environment_variable_t *) calloc(1, sizeof(environment_variable_t));
+
+    CRM_ASSERT(new_entry);
+    *new_entry = (environment_variable_t) {
+        .name = strdup(entry->name),
+        .value = entry->value?strdup(entry->value):NULL
+    };
+    return g_list_prepend(environment_variable_list, new_entry);
```
Comment (kgaillot) [inline crmd/notify.c]: You could simplify this by using copy_environment_variable_entry() instead of new_entry

File: crmd/notify.c [hunk 19]
```diff
@@ -20,154 +20,721 @@
...
+
+        entry.environment_variables =
+            get_envvars_from_cib(notify, 
+                                 entry.environment_variables,
+                                 &envvars);
+
+        crm_notice("Found notification: id=%s, path=%s, timeout=%d, "
+                   "tstamp_format=%s, %d additional environment variables", 
+                   entry.id, entry.path, entry.timeout,
+                   entry.tstamp_format, envvars);
```
Comment (kgaillot) [inline crmd/notify.c]: I think this can be lowered to debug level.

File: xml/crm.dtd (hunk 12 above)
Comment (wenningerk) [inline xml/crm.dtd]: A mystery to observe I have still on my list. Till I had done this alerts wanted to live on top
level instead inside configuration as I remember. I'll observe that again...

File: extra/alerts/pcmk_alert_sample.sh (hunk 16 above)
Comment (wenningerk) [inline extra/alerts/pcmk_alert_sample.sh]: It is moved already. The comments how to use it with ClusterMon are actually still valid and ClusterMon still has options we don't cover with alerts by now (see intro above) alongside of course with disadvantages why we are doing this here. So I would put a note in front that
states how to use it with the new feature...

File: crmd/notify.c (hunk 13 above)
Comment (wenningerk) [inline crmd/notify.c]: Intention was to give the sysadmin insight in the logic when the legacy-notifications are being used and when not. But as the legacy-notifications are anyway a feature that had quite a
short life-span I moved them all to info.

File: crmd/notify.c (hunk 14 above)
Comment (wenningerk) [inline crmd/notify.c]: Was thinking of some kind of priority a recipient might be handled with or
in case of sending out mails the use of a different sender-email or smtp-server
for certain recipients.

File: xml/alerts-2.5.rng (hunk 15 above)
Comment (wenningerk) [inline xml/alerts-2.5.rng]: Well I was searching for something more catchy than "environment_variables" but lacking a 
good idea I tended to calling it what it is.
I was a little afraid of reusing instance_attributes & meta_attributes for something similar but
not quite the same. But you are probably right that the use here is close enough.
But then I would allow anything to be overruled by instance_attributes & meta_attributes in
the recipient sections. Something like a different timeout for a far-away-recipient (in case it
is not just an email-address but a network address that is being contacted) might be
useful as well and if we have the same mimic for all attributes source of confusion is mitigated.

File: crmd/notify.c (hunk 19 above)
Comment (wenningerk) [inline crmd/notify.c]: done - a leftover ...

File: crmd/notify.c (hunk 17 above)
Comment (wenningerk) [inline crmd/notify.c]: true - probably it had once references to allocated stuff or alike ...

File: crmd/notify.c (hunk 18 above)
Comment (wenningerk) [inline crmd/notify.c]: add_dup_envvar was there first ;-)
a lot of room for refactoring in a lot of places - aim of this commit was to nail down the interface

File: xml/alerts-2.5.rng (hunk 15 above)
Comment (wenningerk) [inline xml/alerts-2.5.rng]: ... and of course all the good and the bad coming from defining things more loosely via
the schema. We don't have anything like the resouce-meta-data being used by e.g. pcs
to verify which instance attributes are being added ...

File: xml/crm.dtd (hunk 12 above)
Comment (kgaillot) [inline xml/crm.dtd]: Ah, that made me curious. It looks like RNG_cfg_base in xml/Makefile.am is the key.

File: crmd/notify.c (hunk 14 above)
Comment (kgaillot) [inline crmd/notify.c]: Hmm, I can see the benefit, but I worry the complexity would confuse users (both configuration and troubleshooting). What would a pcs interface look like?

File: xml/alerts-2.5.rng (hunk 15 above)
Comment (kgaillot) [inline xml/alerts-2.5.rng]: That's fine -- we already have that situation with cluster-wide properties.

File: xml/crm.dtd (hunk 12 above)
Comment (wenningerk) [inline xml/crm.dtd]: Effect is gone - no idea why and how - reverted to original.

File: crmd/notify.c (hunk 14 above)
Comment (wenningerk) [inline crmd/notify.c]: pcs: when adding a recipient to an alert then all the attributes that were allowed in the instance_attributes and the meta_attributes section are allowed again.
Implementation is easy because it just calls the code for the alert a 2nd time.
Current state is that just the overruling attributes are in the instance_attributes and meta_attributes section of the recipient section. So it can easily be switched on and
off by just changing the schema.

File: xml/alerts-2.5.rng (hunk 15 above)
Comment (wenningerk) [inline xml/alerts-2.5.rng]: everything that is not optional is in the schema. All the dynamic and optional stuff went
to instance_attributes or meta_attributes.

File: xml/alerts-2.5.rng [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    <element name="alerts">
+      <zeroOrMore>
+        <element name="alert">
+          <attribute name="id"><data type="ID"/></attribute>
+          <optional>
+            <attribute name="description"><text/></attribute>
+          </optional>
+          <!-- path to the script called for alert -->
+          <attribute name="path"><text/></attribute>
+          <ref name="element-resource-extra"/>
```
Comment (kgaillot) [inline xml/alerts-2.5.rng]: typo here and below -- s/resource/alert

**Comment by kgaillot** [comment]
Great work, this was a big project :)

The RNG typo is breaking the build, so fix that, then we can merge.

**Comment by wenningerk** [comment]
Oops embarrassing good that you saw that. Obviously it still finds the stuff from the resources
somehow - didn't expect that. In my tests it worked well ;-)

**Comment by kgaillot** [comment]
Merged :-D

Here are the items needed for the future, in order of importance:

1. Rename publicly exposed *NOTIF* constants in msg_xml.h to *ALERT*.

2. get_meta_attrs_from_cib() should validate user-supplied values (e.g. timeout > 0)

3. for crmd/control.c:config_query_callback(), we should expand crm_read_options() instead of making a separate CIB query

4. update Pacemaker Explained documentation

I'd like to get the above done before the final 1.1.15 release. The rest would be nice if time permits, but can be left for later if necessary:

5. Rename all *notif* files, function names and variables to *alert*. Sample scripts should also be updated to use the new variables (we don't need to reference ClusterMon in the sample scripts anymore -- we should encourage people to use the new interface).

6. Merge high-resolution time handling into existing library time handlers.
